### DISCUSSION TITLE
How do we want to handle external packages that TurfJs depends on like polygon-clipping and d3_geo?

### DISCUSSION THREAD

**Discussion by aardrop** [dartclub/turf_dart#165] [Q&A] [answered]
TurfJS uses a variety of other packages for some GIS math and basic methods, do we want to mirror those as separate projects of merge them as part of this project? 

For instance [TurfJS's buffer method](https://github.com/Turfjs/turf/blob/master/packages/turf-buffer/index.js) requires d3_Geo for the geoAzimuthalEquidistant method. I am looking to start contributing on some of these methods I need for a project but need direction on how the community wants to handle these. 

Let me know what we think and I'll get coding.

**Comment by lukas-h** [comment, marked as answer]
Hi @aardrop, thanks for asking and thanks for the willingness to contribute. We already had this case for a few things. 

https://pub.dev/packages/turf you can look at the dependencies here.

A few soft "rules" for the usage / implementation of these things:

- If there is a really good dart package (that is sure to be maintained the next years and) that contains the functionality we need, we should use this.
- If not:
  - If the functionality makes sense to be exposed it should be a separate small package. It would be good to have these in our organization: https://github.com/dartclub
  - If not: we implement the function internally and don't expose it (either private with an _ underscore or just not exposed via an export)

In this particular case with turf's `buffer` method, let's start just with a private function `_geoAzimuthalEquidistant` and decide later on if we should expose it.

**Comment by aardrop** [comment]
@lukas-h Thanks! 

So I'm about 80% done with a draft of making a dart version of [polygon-clipping](https://www.npmjs.com/package/polygon-clipping) package that turf js depends on. I see that there is a sweepline-intersection package already. There's a sweepline in polygon-clipping. 

I started making the package because it's required by the methods I need from turf and I couldn't find it elsewhere. I can't tell if someone has been working on it internally. So to clarify? should I include this polygon-clipping as a method within the turf package?

Any advice on this? I am hoping to work towards adding difference, buffer, and union with my contributions moving forward.

**Comment by lukas-h** [comment]
@aardrop this comes at the right time!! As we are now also looking to implement the polygon clipping functionality. 

Let's include the polygon-clipping into the turf package and if it makes sense to extract it as a separate package, let's do it afterwards.

If we already have a solid sweepline-intersection implementation, feel free to use that!

CC @jsiedentop

**Reply by lukas-h** [reply]
We haven't started implementing it yet, just some issues and conversations in this repo about it

**Reply by aardrop** [reply]
Great do you want me to push what I have to a forked branch to look at first of should I polish it up for a PR first?

**Reply by aardrop** [reply]
I'm hoping to have the main files done this week and working on final testing and benchmarking

**Reply by lukas-h** [reply]
Please push it to a fork and you can already create a draft Pull Request (where it doesn’t have to be polished 100%)

**Reply by aardrop** [reply]
Will do! I'm heads down on the geom_out file but after that I'll move it and make the draft.

**Reply by lukas-h** [reply]
@jsiedentop and I will help you with code reviews and answer questions

**Comment by aardrop** [comment]
I just pushed a draft; I can't test much to this point, but I will be working on debugging and writing tests this weekend. I added comments to the draft about the general approach, and to the best of my ability, I stuck close to the JS package as possible where it made sense but deviated in instances where stricter classes made things much similar. I added some questions to the draft about the approach and exposing of methods, so if you all have any initial feedback before I focus on testing/debugging functionality, I'd love it.

**Reply by lukas-h** [reply]
I'll give you a review!
Thanks for contributing 👍 👍 👍

**Reply by lukas-h** [reply]
I just left my review here: https://github.com/dartclub/turf_dart/pull/170#pullrequestreview-1886567485
